### DISCUSSION TITLE
Outputting the time derivative of non-linear variable

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#25640] [Q&A General]
Hi all,

I need to output the time derivative of a non-linear variable into another non-linear variable. The type `TimeDerivativeAux` allows me to feed in a non-linear variable but will only output the time derivative to a linear variable. Is there another of computing the time derivative that will let me output to a non-linear variable?

Thank you!

**Comment by GiudGiud** [comment]
Hello

This doesnt seem like a good idea. If Y a non linear variable is simply X_dot, then you are adding a full variable of dofs to the nonlinear system to compute something that you can solve easily on the side. The nonlinear system is bigger and it will cost you

Everywhere you have Y in your equations, just use X_dot instead, which you can compute using `coupledDot("X")` or `adCoupledDot("X")`

Guillaume

**Reply by TheGreatCid** [reply]
That makes sense! Thank you
However, I need to feed this derivative into a ParsedMaterial. `coupledDot` does not make a new coupled variable; it returns a pointer to a value, from my understanding. ParsedMaterial only looks for coupledVariables, which means that I would need to make a new coupled variable to store the time derivatives to feed the dot values in a parsedmaterial. Is this correct?

**Reply by GiudGiud** [reply]
Two options:
- proceed anyway: make a kernel that is just computing test * (Y-X_dot) as the equation for Y. That will do what you want 
- write a new material with the expression you need directly. This is likely the most efficient

**Reply by TheGreatCid** [reply]
Got it, thank you

**Reply by TheGreatCid** [reply]
One more follow up:

>* proceed anyway: make a kernel that is just computing test * (Y-X_dot) as the equation for Y. That will do what you want

I am already using the ADTimeDerivative kernel. My original thought was to just output the `_u_dot[_qp]` variable created by this kernel, but I could not find a way to do so. I just want to double-check as to whether or not this is possible.

**Reply by GiudGiud** [reply]
sure you can use a Reaction kernel and an ADTimeDerivativeKernel to form that equation

**Reply by lindsayad** [reply]
If you don’t have any spatial derivatives in your equation then it doesn’t make sense to add projection error into the problem. I would suggest using the nodal kernel equivalent of time derivative and reaction kernels to avoid this

**Reply by TheGreatCid** [reply]
Got it, thank you both

**Reply by TheGreatCid** [reply]
@GiudGiud
Unfortunately, I am still having some trouble here. I wrote a new kernel but it led to a seg fault when the NL iteration began. I want to make sure I understand everything correctly. Here are the relevant portions of my input file. I have the kernels enacting the physics for `d`, while I have an extra kernel to output the derivative of `d` with respect to time, which is written to the variable `ddot`

```
[Variables]
  [d]
  []
  [ddot]
  []
[]

[Kernels]
  [Phys1]
    type = CoefTimeDerivative
    variable = d
  [phy2]
   type = ADDiffusion
   variable = d
  [phy3]
    type = ADReaction
    variable = d 
  []
   [ddot]
    type = ADTimeDerivativeCustom
    variable = ddot
    deriv_var = d
  []
[]
```

Here is the relevant section of the kernel I wrote. I am replicating the _test(Y-X_dot) form you suggested. 
Are you able to find any issues with this implementation? 

```InputParameters
ADTimeDerivativeCustom::validParams()
{
  InputParameters params = ADKernelValue::validParams();
  params.addClassDescription("outputting the time derivative");
  params.addRequiredCoupledVar("deriv_var", "variable to take time derivative of");

  return params;
}

ADTimeDerivativeCustom::ADTimeDerivativeCustom(const InputParameters & parameters)
  : ADKernelValue(parameters), _d_dot(adCoupledDot("deriv_var"))
{
}

ADReal
ADTimeDerivativeCustom::precomputeQpResidual()
{
  return _test[_i][_qp] * (_u[_qp] - _d_dot[_qp]);
}
```

**Reply by lindsayad** [reply]
Try compiling in dbg mode and run in a debugger. That will show you where the segmentation fault is coming from

**Reply by TheGreatCid** [reply]
Ah I was afraid you would say that. I need to figure out why I get compile errors when trying to make the dbg executable. But that's a whole other topic

**Reply by TheGreatCid** [reply]
Ok finally got things fixed
The reason for the seg vault is a bit vague 
```
Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 0 size: 0)
at /home/det12/projects/DavidRac/raccoon/moose/framework/build/header_symlinks/MooseArray.h, line 275
```

If you have an insight into this it would be much appreciated. 

edit: ah this is occurring in the fix point solve. This input file is an subapp, where it iterates on `d` to minimize. I'm guessing that throwing in another nl variable is causing issues

**Reply by lindsayad** [reply]
A full backtrace would be helpful. My guess is that `_d_dot` is not getting correctly sized for some reason

**Reply by lindsayad** [reply]
We would need an input file and the code you are using to track down the problem if the issue is `_d_dot` not being correctly sized

**Reply by TheGreatCid** [reply]
[files.zip](https://github.com/idaholab/moose/files/12785640/files.zip)
Here is the input file as well as the custom kernel I created. 

Additionally here is the full back trace

```
Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 0 size: 0)
at /home/det12/projects/DavidRac/raccoon/moose/framework/build/header_symlinks/MooseArray.h, line 275
Stack frames: 30
0: libMesh::print_trace(std::ostream&)
1: /home/det12/projects/DavidRac/raccoon/lib/libraccoon-dbg.so.0(+0xcb985) [0x7f9e066de985]
2: ADTimeDerivativeCustom::precomputeQpResidual()
3: ADKernelValueTempl<double>::computeResidual()
4: ComputeResidualThread::compute(ResidualObject&)
5: NonlinearThread::compute(KernelBase&)
6: NonlinearThread::computeOnElement()
7: NonlinearThread::onElement(libMesh::Elem const*)
8: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
9: NonlinearThread::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
10: /home/det12/projects/DavidRac/raccoon/moose/framework/libmoose-dbg.so.0(+0x15c7f48) [0x7f9e014baf48]
11: NonlinearSystemBase::computeResidualInternal(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
12: NonlinearSystemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
13: FEProblemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
14: FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
15: FEProblemBase::computeResidual(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, unsigned int)
16: FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&)
17: NonlinearSystem::solve()
18: FEProblemBase::solve(unsigned int)
19: FEProblemSolve::solve()
20: FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&)
21: FixedPointSolve::solve()
22: TimeStepper::step()
23: Transient::takeStep(double)
24: Transient::execute()
25: MooseApp::executeExecutioner()
26: MooseApp::run()
27: main
28: __libc_start_main
29: ./raccoon-dbg(+0x3079) [0x55a5c36fc079]
```

**Reply by TheGreatCid** [reply]
Ok I fixed the issue. My custom class was inherited from `kernelvalue.h`. I needed to just inherit from `ADKernel.h`. I appreciate the help

**Reply by lindsayad** [reply]
I'm glad you found a solution! I'm still going to investigate this though as this is not something a user should have run into

**Reply by lindsayad** [reply]
You would eventually run into issues with out of bounds access into `_test[_i][_qp]` while inheriting from `ADKernelValue` (which does the test multiplication for you as an optimization), so it's good that you switched to `ADKernel`.

However, the out of bounds access at index 0 is because you did not store `_d_dot` as a reference in your header file. You copied it by value. So you have to make sure in your `.h` file that `_d_dot` looks like 
```c++
  const ADVariableValue & _d_dot;
```
and not
```c++
  const ADVariableValue _d_dot;
```

**Reply by TheGreatCid** [reply]
Noted; thank you for looking into this!
